Documentation/op-guide/performance.md

```diff
@@ -68,3 +68,112 @@ benchmark --endpoints=${HOST_1},${HOST_2},${HOST_3} --conns=100 --clients=1000 \
+
+Sample fio command random read/write 16k blocks: * Note this will generate a 4GB file.
+
+```sh
+fio --randrepeat=1 \
+  --ioengine=libaio \
+  --direct=1 \
+  --gtod_reduce=1 \
+  --name=etcd-disk-io-test \
+  --filename=etcd_read_write.io \
+  --bs=16k --iodepth=64 --size=4G \
+  --readwrite=randrw --rwmixread=75
+```
+
+Example output of SSD drive:
+
+```sh
+etcd-disk-io-test: (g=0): rw=randrw, bs=(R) 4096B-4096B, (W) 4096B-4096B, (T) 4096B-4096B, ioengine=libaio, iodepth=64
+fio-3.7
+Starting 1 process
+etcd-disk-io-test: Laying out IO file (1 file / 4096MiB)
+Jobs: 1 (f=1): [m(1)][100.0%][r=362MiB/s,w=119MiB/s][r=92.6k,w=30.6k IOPS][eta 00m:00s]
+etcd-disk-io-test: (groupid=0, jobs=1): err= 0: pid=7019: Fri Mar 22 09:57:17 2019
+   read: IOPS=90.1k, BW=352MiB/s (369MB/s)(3070MiB/8722msec)
+   bw (  KiB/s): min=323568, max=391608, per=100.00%, avg=361922.35, stdev=17944.92, samples=17
+   iops        : min=80892, max=97902, avg=90480.59, stdev=4486.23, samples=17
+  write: IOPS=30.1k, BW=118MiB/s (123MB/s)(1026MiB/8722msec)
+   bw (  KiB/s): min=106504, max=130680, per=100.00%, avg=120970.35, stdev=6126.94, samples=17
+   iops        : min=26626, max=32670, avg=30242.59, stdev=1531.74, samples=17
+  cpu          : usr=12.53%, sys=52.90%, ctx=135891, majf=0, minf=7
+  IO depths    : 1=0.1%, 2=0.1%, 4=0.1%, 8=0.1%, 16=0.1%, 32=0.1%, >=64=100.0%
+     submit    : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.0%, >=64=0.0%
+     complete  : 0=0.0%, 4=100.0%, 8=0.0%, 16=0.0%, 32=0.0%, 64=0.1%, >=64=0.0%
+     issued rwts: total=785920,262656,0,0 short=0,0,0,0 dropped=0,0,0,0
+     latency   : target=0, window=0, percentile=100.00%, depth=64
+
+Run status group 0 (all jobs):
+   READ: bw=352MiB/s (369MB/s), 352MiB/s-352MiB/s (369MB/s-369MB/s), io=3070MiB (3219MB), run=8722-8722msec
+  WRITE: bw=118MiB/s (123MB/s), 118MiB/s-118MiB/s (123MB/s-123MB/s), io=1026MiB (1076MB), run=8722-8722msec
+
+Disk stats (read/write):
+    dm-4: ios=767338/256593, merge=0/0, ticks=435534/40456, in_queue=476977, util=98.96%, aggrios=395222/131476, aggrmerge=0/0, aggrticks=225096/20908, aggrin_queue=246457, aggrutil=98.64%
+    dm-1: ios=790444/262880, merge=0/0, ticks=450192/41589, in_queue=492687, util=98.64%, aggrios=790444/262680, aggrmerge=0/200, aggrticks=440412/37872, aggrin_queue=482628, aggrutil=100.00%
+  nvme0n1: ios=790444/262680, merge=0/200, ticks=440412/37872, in_queue=482628, util=100.00%
+    dm-0: ios=0/72, merge=0/0, ticks=0/227, in_queue=227, util=0.66%, aggrios=0/36, aggrmerge=0/36, aggrticks=0/119, aggrin_queue=389, aggrutil=3.78%
+  nvme1n1: ios=0/36, merge=0/36, ticks=0/119, in_queue=389, util=3.78%
+```
+
+Sample fio command seqential write 32k blocks: * Note this will generate a 4GB file.
+
+```sh
+fio --randrepeat=1 \
```

Comment:
I suggest using `--ioengine=sync` and adding `--fdatasync=1`. 

Our specific use case was measuring fdatasync latency when writing to Etcd WALs. @MikeSpreitzer did a thorough analysis of the system calls Etcd uses to write to WALs using [strace](https://linux.die.net/man/1/strace). He analyzed both Etcd 3.2.24 (the version we're using) and Etcd 3.3.12 (latest). No relevant difference was found. There's always a [write](https://linux.die.net/man/2/write) immediately followed by an [fdatasync](https://linux.die.net/man/2/fdatasync).  Here's an excerpt of strace output (8 is the fd of the WAL file): 
```
21:23:09.894911 write(8, ".\0\0\0\0\0\0\202\10\2\20\361\223\255\266\6\32$\10\0\20\10\30\26\"\34\"\r\n\3fo"..., 96) = 96 <0.0000\
30>
21:23:09.894966 fdatasync(8)            = 0 <0.000014>
```

As for fio, using `--ioengine=sync` and `--fdatasync=1` achieves the same result: disk is written to with a [write](https://linux.die.net/man/2/write) immediately followed by an [fdatasync](https://linux.die.net/man/2/fdatasync).  Again, this has been checked by launching fio with strace. Here's an excerpt of strace ouput:
```
14:06:38.102814 write(3, "xq\235\2178\267i\217/\2567Y\30\357\230\22\305u\0174G\361O\16\270n\237+!dI\24"..., 2200) = 2200 <0.000053>
14:06:38.102926 fdatasync(3)            = 0 <0.010523>
```
Notice that this way you get disk I/O similar to the way Etcd writes to WALs, but Etcd does other disk I/O as well and we did not look at that. All tests have been run on Ubuntu 18.04.